### ISSUE TITLE
missing HttpRequest fields are erroneously set to 'nil' rather than left unset.

### ISSUE THREAD

**Issue by mr-salty** [GoogleCloudPlatform/fluent-plugin-google-cloud#84]
The code that constructs an HttpRequest sets fields that are not present in the input to 'nil' rather than leaving them unset.  This causes those fields to appear in the request sent to the API with null values rather than being omitted as they should be (the [serialization does make a distinction](https://github.com/google/google-api-ruby-client/blob/1ee88320e91deccf4e69fe589b4959046dcc752e/lib/google/apis/core/json_representation.rb#L49)).

so instead of:

`output.request_method = input.delete('requestMethod')`

we should have

`output.request_method = input.delete('requestMethod') if input.key?('requestMethod')`

(FYI, to look at the actual requests I added a line to lib/google/apis/core/api_command.rb):

```
self.body = request_representation.new(request_object).to_json(skip_undefined: true)
File.open('/tmp/requests', 'a') { |file| file.write(self.body); file.write("\n") }
```

**Comment by christopherL91** [comment]
What is the status on this issue? I have the same problem as found in #85. I use nginx and just output the log to stdout which in turn is captured by stackdriver on GCE.

```json
{
 metadata: {…}   
 textPayload: "2016-12-06 17:16:28 +0000 [warn]: Dropping 1 log message(s) error_class="Google::Apis::ClientError" error="badRequest: Invalid JSON payload received. Unknown name \"validated_with_origin_server\" at 'entries[0].http_request': Cannot find field."
"   
 insertId: "1thtsqzekp5jl"   
 log: "fluentd-cloud-logging"   
}
```

**Comment by ged15** [comment]
Hello!
Any progress regarding this issue?

**Comment by igorpeshansky** [comment]
As far as I know, the latest version of the gem (0.6.0) already does this. @qingling128, could you please verify that and resolve this issue?

**Comment by qingling128** [comment]
Looking.

**Comment by qingling128** [comment]
For gRPC path (if you set `use_grpc true` in the config), this has already been fixed: If a field is `null` or absent from the input, it will be left unset in the constructed proto (Setting it to `null` does not apply to the gRPC path since the proto has these fields defined as strings or integers).

For the non-gRPC path (if you set `use_grpc false` in the config or leave the `use_grpc` flag unset), the current behavior is: If a field is `null` or absent from the input, it will be set as `null` in the constructed json. (Note that even if we set a field to `null` in the json, when it reaches the Logging API, the field would end up being unset during the conversion to proto. So the end result is the same between the gRPC and non-gRPC paths.)

I just submitted https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/pull/120 to modify the behavior in the non-gRPC path: for fields absent from the input, we will leave them unset in the constructed json as well.

**Comment by qingling128** [comment]
Slight changes to the implementation: after some discussion, we decided to leave both `null` and absent fields unset in both the gRPC and non-gRPC paths, so the behavior is consistent. Because the end result after converting to protos would be the same regardless of whether we leave it unset or set it null.

**Comment by qingling128** [comment]
Fix in. Will be released in `v0.6.1`.
